File: tools/extract-encoding-table.m
// Produces the definitive value-encoding table for .shortcut files by asking
// WorkflowKit's own classes to serialize:
//   - every WFVariable kind (action output, named variable, shortcut input, ...)
//   - a WFVariableString with embedded variables (the WFTextTokenString form)
//   - every parameter state class, constructed with a plain value and with a variable
//   - the 15 icon palette colors as WFWorkflowIcon reports them
//   - the operator codes each conditional subject type supports
//
//   xcrun clang -fobjc-arc -fobjc-exceptions -framework Foundation -o /tmp/enc extract-encoding-table.m
//   /tmp/enc out.json [data/parameter-encodings.json]
// The optional second argument supplies the state class names (from
// extract-parameter-encodings.m); without it a built-in list is used. Scanning every
// class in the process is avoided on purpose: +initialize on unrelated classes crashes.
#import <Foundation/Foundation.h>
#import <objc/runtime.h>
#import <objc/message.h>
#include <dlfcn.h>

static id c0(id t, const char* s) { return ((id (*)(id, SEL))objc_msgSend)(t, sel_registerName(s)); }
static id c1(id t, const char* s, id a) { return ((id (*)(id, SEL, id))objc_msgSend)(t, sel_registerName(s), a); }
static id plain(id v, int d) {
  if (!v) return [NSNull null];
  if (d > 8) return [v description];
  if ([v isKindOfClass:[NSString class]] || [v isKindOfClass:[NSNumber class]]) return v;
  if ([v isKindOfClass:[NSDictionary class]]) { NSMutableDictionary* o = [NSMutableDictionary new]; for (id k in v) o[[k description]] = plain(v[k], d + 1); return o; }
  if ([v isKindOfClass:[NSArray class]]) { NSMutableArray* a = [NSMutableArray new]; for (id x in v) [a addObject:plain(x, d + 1)]; return a; }
  return [NSString stringWithFormat:@"<%s>", class_getName([v class])];
}
static id attempt(id (^block)(void)) { @try { return block(); } @catch (NSException* e) { return [NSString stringWithFormat:@"threw: %@", e.reason]; } }
static id serialized(id state) { return state ? plain(c0(state, "serializedRepresentation"), 0) : @"nil"; }

int main(int argc, char** argv) {
  if (argc < 2) { fprintf(stderr, "usage: %s out.json\n", argv[0]); return 2; }
  if (!dlopen("/System/Library/PrivateFrameworks/WorkflowKit.framework/WorkflowKit", RTLD_NOW)) { fprintf(stderr, "dlopen failed\n"); return 1; }
  NSMutableDictionary* out = [NSMutableDictionary new];

  // Variables
  id named = ((id (*)(id, SEL, id, id, id))objc_msgSend)([objc_getClass("WFUserDefinedVariable") alloc], sel_registerName("initWithName:variableProvider:aggrandizements:"), @"MyVar", nil, nil);
  id output = ((id (*)(id, SEL, id, id, id, id))objc_msgSend)([objc_getClass("WFActionOutputVariable") alloc], sel_registerName("initWithOutputUUID:outputName:variableProvider:aggrandizements:"), @"11111111-2222-3333-4444-555555555555", @"Stored Content", nil, nil);
  NSMutableDictionary* vars = [NSMutableDictionary new];
  vars[@"WFUserDefinedVariable(MyVar)"] = plain(c0(named, "serializedRepresentation"), 0);
  vars[@"WFActionOutputVariable"] = plain(c0(output, "serializedRepresentation"), 0);
  vars[@"WFShortcutInputVariable"] = plain(c0(((id (*)(id, SEL, id, id))objc_msgSend)([objc_getClass("WFShortcutInputVariable") alloc], sel_registerName("initWithVariableProvider:aggrandizements:"), nil, nil), "serializedRepresentation"), 0);
  for (NSString* cls in @[@"WFClipboardVariable", @"WFCurrentDateVariable", @"WFDeviceDetailsVariable", @"WFCurrentAppVariable"])
    vars[cls] = plain(c0(c1([objc_getClass(cls.UTF8String) alloc], "initWithAggrandizements:", nil), "serializedRepresentation"), 0);
  vars[@"WFAskEachTimeVariable(prompt)"] = plain(c0(c1([objc_getClass("WFAskEachTimeVariable") alloc], "initWithPrompt:", @"Which one?"), "serializedRepresentation"), 0);
  out[@"variables"] = vars;

  // Text with embedded variables
  id vs = c1([objc_getClass("WFVariableString") alloc], "initWithStringsAndVariables:", @[@"Hello ", named, @"!"]);
  out[@"variableString(Hello <MyVar>!)"] = plain(c0(vs, "serializedRepresentation"), 0);

  // State classes: plain value and variable forms
  NSMutableDictionary* states = [NSMutableDictionary new];
  NSMutableSet* stateNames = [NSMutableSet setWithArray:@[@"WFStringSubstitutableState", @"WFBooleanSubstitutableState", @"WFNumberSubstitutableState", @"WFNumberStringSubstitutableState", @"WFVariableStringParameterState", @"WFVariableParameterState", @"WFURLStringParameterState", @"WFDateFieldParameterState", @"WFStringParameterState", @"WFNumberParameterState", @"WFDictionaryParameterState", @"WFAppDescriptorParameterState", @"WFColorParameterState", @"WFFileParameterState", @"WFLocationParameterState", @"WFINObjectSubstitutableState", @"WFCalendarSubstitutableState", @"WFWorkflowParameterState", @"WFQuantityParameterState", @"WFContactFieldEntry", @"WFFontParameterState", @"WFNSUnitSubstitutableState", @"WFIntentDescriptorParameterState", @"WFMediaItemState"]];
  if (argc > 2) { NSDictionary* enc = [NSJSONSerialization JSONObjectWithData:[NSData dataWithContentsOfFile:[NSString stringWithUTF8String:argv[2]]] options:0 error:nil];
    for (NSDictionary* e in [enc[@"parameterClasses"] allValues]) { id sc = e[@"stateClass"]; if ([sc isKindOfClass:[NSString class]] && ![sc hasPrefix:@"_TtGC"]) [stateNames addObject:sc]; } }
  for (NSString* cn0 in [[stateNames allObjects] sortedArrayUsingSelector:@selector(compare:)]) {
    const char* cn = cn0.UTF8String; Class c = objc_getClass(cn); if (!c) continue;
    BOOL relevant = [c instancesRespondToSelector:sel_registerName("serializedRepresentation")] && (
      [c instancesRespondToSelector:sel_registerName("initWithValue:")] || [c instancesRespondToSelector:sel_registerName("initWithVariable:")] ||
      [c instancesRespondToSelector:sel_registerName("initWithNumber:")] || [c instancesRespondToSelector:sel_registerName("initWithString:")] ||
      [c instancesRespondToSelector:sel_registerName("initWithVariableString:")] || [c instancesRespondToSelector:sel_registerName("initWithKeyValuePairs:")]);
    if (!relevant) continue;
    NSMutableDictionary* e = [NSMutableDictionary new];
    if ([c instancesRespondToSelector:sel_registerName("initWithValue:")]) {
      NSDictionary* cands = @{@"value:string": @"hello", @"value:number": @5, @"value:bool": @YES};
      for (NSString* k in cands) { id r = attempt(^{ return serialized(c1([c alloc], "initWithValue:", cands[k])); }); if (!([r isKindOfClass:[NSString class]] && [r hasPrefix:@"threw"])) e[k] = r; }
    }
    if ([c instancesRespondToSelector:sel_registerName("initWithNumber:")]) e[@"number:5"] = attempt(^{ return serialized(c1([c alloc], "initWithNumber:", @5)); });
    if ([c instancesRespondToSelector:sel_registerName("initWithBoolValue:")]) e[@"bool:YES"] = attempt(^{ return serialized(((id (*)(id, SEL, BOOL))objc_msgSend)([c alloc], sel_registerName("initWithBoolValue:"), YES)); });
    if ([c instancesRespondToSelector:sel_registerName("initWithString:")]) e[@"string:hello"] = attempt(^{ return serialized(c1([c alloc], "initWithString:", @"hello")); });
    if ([c instancesRespondToSelector:sel_registerName("initWithVariableString:")]) e[@"variableString:Hello <MyVar>!"] = attempt(^{ return serialized(c1([c alloc], "initWithVariableString:", vs)); });
    if ([c instancesRespondToSelector:sel_registerName("initWithVariable:")]) {
      e[@"variable:MyVar"] = attempt(^{ return serialized(c1([c alloc], "initWithVariable:", named)); });
      e[@"variable:ActionOutput"] = attempt(^{ return serialized(c1([c alloc], "initWithVariable:", output)); });
    }
    if ([c instancesRespondToSelector:sel_registerName("initWithKeyValuePairs:")]) e[@"keyValuePairs:[]"] = attempt(^{ return serialized(c1([c alloc], "initWithKeyValuePairs:", @[])); });
    if (e.count) states[[NSString stringWithUTF8String:cn]] = e;
  }
  out[@"stateClasses"] = states;

  // Icon palette
  NSMutableArray* palette = [NSMutableArray new]; Class icon = objc_getClass("WFWorkflowIcon");
  for (unsigned long i = 0; i < 15; i++) {
    id ic = ((id (*)(id, SEL, unsigned long, unsigned short))objc_msgSend)([icon alloc], sel_registerName("initWithPaletteColor:glyphCharacter:"), i, 61440);
    long long v = ((long long (*)(id, SEL))objc_msgSend)(ic, sel_registerName("backgroundColorValue"));
    [palette addObject:@{@"palette": @(i), @"WFWorkflowIconStartColor": @((unsigned int)(v & 0xffffffff)), @"signed": @(v), @"color": [c0(ic, "backgroundColor") description] ?: @""}];
  }
  out[@"iconPalette"] = palette;
  out[@"defaultGlyphCharacter"] = @(((unsigned long (*)(id, SEL))objc_msgSend)(icon, sel_registerName("defaultGlyphCharacter")));

  // Conditional subject operator codes
  NSMutableDictionary* ops = [NSMutableDictionary new];
  for (NSString* cn0 in @[@"WFConditionalSubjectParameterState", @"WFVariableConditionalSubjectState", @"WFHomeAccessoryConditionalSubjectState"]) { Class c = objc_getClass(cn0.UTF8String); if (!c) continue;
    id st = attempt(^{ return [[c alloc] init]; }); id o = [st isKindOfClass:[NSString class]] ? st : attempt(^{ return plain(c0(st, "supportedComparisonOperators"), 0); });
    ops[cn0] = @{@"subjectType": plain(attempt(^{ return c0(c, "subjectType"); }), 0), @"operators": o ?: [NSNull null]}; }
  out[@"conditionalSubjectOperators"] = ops;

  NSData* json = [NSJSONSerialization dataWithJSONObject:out options:NSJSONWritingPrettyPrinted | NSJSONWritingSortedKeys error:nil];
  [json writeToFile:[NSString stringWithUTF8String:argv[1]] atomically:YES];
  fprintf(stderr, "%lu state classes, %lu variables -> %s\n", (unsigned long)states.count, (unsigned long)vars.count, argv[1]);
  return 0;
}

File: tools/extract-encoding-table.js
// Produces the definitive value-encoding table for .shortcut files by asking
// WorkflowKit's own classes to serialize:
//   - every WFVariable kind (action output, named variable, shortcut input, ...)
//   - a WFVariableString with embedded variables (the WFTextTokenString form)
//   - every parameter state class, constructed with a plain value and with a variable
//   - the 15 icon palette colors as WFWorkflowIcon reports them
//   - the operator codes each conditional subject type supports
// See jxa-prelude.js for why this is JXA.
//
//   cat tools/jxa-prelude.js tools/extract-encoding-table.js > /tmp/x.js
//   osascript -l JavaScript /tmp/x.js data/encoding-table.json [data/parameter-encodings.json]
// The optional second argument supplies the state class names (from
// extract-parameter-encodings.js); without it a built-in list is used. Scanning every
// class in the process is avoided on purpose: +initialize on unrelated classes crashes.
function attempt(fn) { try { return fn(); } catch (e) { return "threw: " + e.message; } }
function serialized(state) { return isNil(state) ? "nil" : plain(state.serializedRepresentation, 0); }
function run(argv) {
  if (argv.length < 1) { throw new Error("usage: extract-encoding-table.js out.json [parameter-encodings.json]"); }
  loadEngine();
  const out = {};
  // Variables
  const named = make("WFUserDefinedVariable", "initWithName:variableProvider:aggrandizements:", "MyVar", null, null);
  const output = make("WFActionOutputVariable", "initWithOutputUUID:outputName:variableProvider:aggrandizements:", "11111111-2222-3333-4444-555555555555", "Stored Content", null, null);
  const vars = {};
  vars["WFUserDefinedVariable(MyVar)"] = plain(named.serializedRepresentation, 0);
  vars["WFActionOutputVariable"] = plain(output.serializedRepresentation, 0);
  vars["WFShortcutInputVariable"] = plain(make("WFShortcutInputVariable", "initWithVariableProvider:aggrandizements:", null, null).serializedRepresentation, 0);
  for (const cn of ["WFClipboardVariable", "WFCurrentDateVariable", "WFDeviceDetailsVariable", "WFCurrentAppVariable"])
    vars[cn] = plain(make(cn, "initWithAggrandizements:", null).serializedRepresentation, 0);
  vars["WFAskEachTimeVariable(prompt)"] = plain(make("WFAskEachTimeVariable", "initWithPrompt:", "Which one?").serializedRepresentation, 0);
  out.variables = vars;

  // Text with embedded variables
  const vs = make("WFVariableString", "initWithStringsAndVariables:", $(["Hello ", named, "!"]));
  out["variableString(Hello <MyVar>!)"] = plain(vs.serializedRepresentation, 0);

  // State classes: plain value and variable forms
  const states = {};
  const stateNames = new Set(["WFStringSubstitutableState", "WFBooleanSubstitutableState", "WFNumberSubstitutableState", "WFNumberStringSubstitutableState", "WFVariableStringParameterState", "WFVariableParameterState", "WFURLStringParameterState", "WFDateFieldParameterState", "WFStringParameterState", "WFNumberParameterState", "WFDictionaryParameterState", "WFAppDescriptorParameterState", "WFColorParameterState", "WFFileParameterState", "WFLocationParameterState", "WFINObjectSubstitutableState", "WFCalendarSubstitutableState", "WFWorkflowParameterState", "WFQuantityParameterState", "WFContactFieldEntry", "WFFontParameterState", "WFNSUnitSubstitutableState", "WFIntentDescriptorParameterState", "WFMediaItemState"]);
  if (argv.length > 1) {
    const enc = readJSON(argv[1]);
    for (const e of Object.values(enc.parameterClasses)) if (typeof e.stateClass === "string" && !e.stateClass.startsWith("_TtGC")) stateNames.add(e.stateClass);
  }
  const respondsTo = (c, sel) => c.instancesRespondToSelector(sel);
  for (const cn of [...stateNames].sort()) {
    if (!hasClass(cn)) continue;
    const c = klass(cn);
    const relevant = respondsTo(c, "serializedRepresentation") && (
      respondsTo(c, "initWithValue:") || respondsTo(c, "initWithVariable:") || respondsTo(c, "initWithNumber:") || respondsTo(c, "initWithString:") ||
      respondsTo(c, "initWithVariableString:") || respondsTo(c, "initWithKeyValuePairs:"));
    if (!relevant) continue;
    const e = {};
    if (respondsTo(c, "initWithValue:")) {
      const cands = { "value:string": "hello", "value:number": 5, "value:bool": true };
      for (const k of Object.keys(cands)) { const r = attempt(() => serialized(make(cn, "initWithValue:", cands[k]))); if (!(typeof r === "string" && r.startsWith("threw"))) e[k] = r; }
    }
    if (respondsTo(c, "initWithNumber:")) e["number:5"] = attempt(() => serialized(make(cn, "initWithNumber:", 5)));
    if (respondsTo(c, "initWithBoolValue:")) e["bool:YES"] = attempt(() => serialized(makeBool(cn, "initWithBoolValue:", true)));
    if (respondsTo(c, "initWithString:")) e["string:hello"] = attempt(() => serialized(make(cn, "initWithString:", "hello")));
    if (respondsTo(c, "initWithVariableString:")) e["variableString:Hello <MyVar>!"] = attempt(() => serialized(make(cn, "initWithVariableString:", vs)));
    if (respondsTo(c, "initWithVariable:")) {
      e["variable:MyVar"] = attempt(() => serialized(make(cn, "initWithVariable:", named)));
      e["variable:ActionOutput"] = attempt(() => serialized(make(cn, "initWithVariable:", output)));
    }
    if (respondsTo(c, "initWithKeyValuePairs:")) e["keyValuePairs:[]"] = attempt(() => serialized(make(cn, "initWithKeyValuePairs:", $([]))));
    if (Object.keys(e).length) states[cn] = e;
  }
  out.stateClasses = states;

  // Icon palette
  const palette = []; const icon = klass("WFWorkflowIcon");
  for (let i = 0; i < 15; i++) {
    const ic = makeUInts("WFWorkflowIcon", "initWithPaletteColor:glyphCharacter:", i, 61440);
    const v = Number(ic.backgroundColorValue); // the bridge returns 64-bit integers as strings
    palette.push({ palette: i, WFWorkflowIconStartColor: v >>> 0, signed: v, color: str(ic.backgroundColor.description) || "" });
  }
  out.iconPalette = palette;
  out.defaultGlyphCharacter = icon.defaultGlyphCharacter;

  // Conditional subject operator codes
  const ops = {};
  for (const cn of ["WFConditionalSubjectParameterState", "WFVariableConditionalSubjectState", "WFHomeAccessoryConditionalSubjectState"]) {
    if (!hasClass(cn)) continue;
    const c = klass(cn);
    const st = attempt(() => make(cn, "init"));
    const o = typeof st === "string" ? st : attempt(() => plain(st.supportedComparisonOperators, 0));
    ops[cn] = { subjectType: plain(attempt(() => c.subjectType), 0), operators: o === undefined ? null : o };
  }
  out.conditionalSubjectOperators = ops;

  writeJSON(argv[0], out);
  console.log(Object.keys(states).length + " state classes, " + Object.keys(vars).length + " variables -> " + argv[0]);
}
